### PR TITLE
Proposed fix for embedded structs are not honoured

### PR DESCRIPTION
**Pull request by stromsvag**
This is a proposed fix for #376 .

I added [a comment](https://github.com/cuelang/cue/issues/376#issuecomment-715940879) to #376 with some questions. Would be great if I could get some feedback on that as well as my proposed fix. My experience with cue is very limited.

### CODE REVIEW

File: encoding/gocode/gocodec/codec_test.go [hunk 1]
```diff
@@ -263,10 +263,27 @@ func TestDecode(t *testing.T) {
...
+					A string
+					T
+				}
+				return S{}
+			}(),
+			want: `{
+	A: ""
+	B: 0
+}`,
+		},
```
Comment (mpvl) [inline encoding/gocode/gocodec/codec_test.go:285]: The purpose of this change is to let the CUE encoding mimic Go's encoding.

There should also be a test case where an embedded struct is assigned a name. For instance:
```
			type S struct {
					A string
					T `json:"t"`
				}
```
In that case, the type should not be embedded but named that way.

File: internal/core/convert/go.go [hunk 2]
```diff
@@ -430,20 +430,30 @@ func convertRec(ctx *adt.OpContext, nilIsTop bool, x interface{}) adt.Value {
...
 				if name == "-" {
 					continue
 				}
-				f := ctx.StringLabel(name)
-				obj.Decls = append(obj.Decls, &adt.Field{Label: f, Value: sub})
-				arc, ok := sub.(*adt.Vertex)
-				if ok {
-					arc.Label = f
+				if sf.Anonymous {
+					obj.Decls = append(obj.Decls, sub)
```
Comment (mpvl) [inline internal/core/convert/go.go:438]: I would expect you'ld either embed sub or add its arcs below, but not both. Adding the arcs is probably nicer.

**Review by mpvl — CHANGES_REQUESTED** [review]

File: internal/core/convert/go.go [hunk 3]
```diff
@@ -405,18 +405,18 @@ func convertRec(ctx *adt.OpContext, nilIsTop bool, x interface{}) adt.Value {
 
 			t := value.Type()
 			for i := 0; i < value.NumField(); i++ {
-				t := t.Field(i)
-				if t.PkgPath != "" {
+				sf := t.Field(i)
```
[the remark this answers is not in the document]
Comment (stromsvag) [reply internal/core/convert/go.go:408]: The only reason is that I found it a bit confusing to name it the same. I find `t` a great variable name for "type", but perhaps not the best for a struct field. Also, my IDE reported it as being shadowed. I am fine with reverting this change.

File: internal/core/convert/go.go (hunk 2 above)
Comment (stromsvag) [reply internal/core/convert/go.go:438]: I currently have to do both in order to pass testing. I am still very new to this so I might have fundamentally misunderstood something.

File: internal/core/convert/go.go [hunk 4]
```diff
@@ -430,20 +430,30 @@ func convertRec(ctx *adt.OpContext, nilIsTop bool, x interface{}) adt.Value {
...
+				name := getName(&sf)
 				if name == "-" {
 					continue
 				}
-				f := ctx.StringLabel(name)
-				obj.Decls = append(obj.Decls, &adt.Field{Label: f, Value: sub})
-				arc, ok := sub.(*adt.Vertex)
-				if ok {
-					arc.Label = f
+				if sf.Anonymous {
```
[the remark this answers is not in the document]
Comment (stromsvag) [reply internal/core/convert/go.go:437]: I don't think name will ever be empty. However, I could modify `getName` to return empty name if struct field is anonymous and not tagged. The condition could then be `if sf.Anonymous && name == ""`. That should fulfill the case you commented in tests.

File: encoding/gocode/gocodec/codec_test.go (hunk 1 above)
Comment (stromsvag) [reply encoding/gocode/gocodec/codec_test.go]: I added a proposed solution for this. Let me know what you think.

File: encoding/gocode/gocodec/codec_test.go [hunk 5]
```diff
@@ -263,10 +263,27 @@ func TestDecode(t *testing.T) {
 	testCases := []struct {
 		in   interface{}
 		want string
-	}{{
-		in:   "str",
-		want: `"str"`,
-	}}
+	}{
+		{
```
[the remark this answers is not in the document]
Comment (stromsvag) [reply encoding/gocode/gocodec/codec_test.go]: Done :) I will start following this convention.

File: internal/core/convert/go.go (hunk 2 above)
Comment (stromsvag) [reply internal/core/convert/go.go]: I took a stab at a change that I hope addresses your comment. See 326ff87.

File: internal/core/convert/go.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (stromsvag) [reply internal/core/convert/go.go]: I decreased nesting by adding continue.

**Review by mpvl — COMMENTED** [review]
Moving it go Gerrit.

**Pull request closed without merging**